### PR TITLE
Filename extensions: Allow specifying wildcards

### LINKED ISSUE #1561: Series: Open with Placeholder for File-Ending/Backend
I realized that in many situations, generalized post-processing scripts, CI, etc., it would be cool if we could open a series of the kind
```C++
auto s = openPMD::Series("samples/git-sample/data%T.*", openPMD::Access::READ_ONLY);
```
or
```C++
auto s = openPMD::Series("samples/git-sample/data%T.%B", openPMD::Access::READ_ONLY);
```

without having to specify the backend. The logic would just match the first backend that fits the pattern.

In openPMD-viewer, we approach this by just specifying the directory in which files are placed, which has the downside that there cannot be multiple series in the same directory.

What do you think, @franzpoeschel ? :)

### PR DESCRIPTION
**Pull request by franzpoeschel**
Close #1561

- [x] Merge #1557 first

Diff: https://github.com/franzpoeschel/openPMD-api/compare/refactor-extract-bufferedactions..topic-wildcard-extension

Usage idea:

```
> bin/openpmd-pipe --infile ../samples/git-sample/3d-bp4/example-3d-bp4_%T.%E --outfile out.%E                                                              
Opening data source                                                                                                                                         
Opening data sink                                                                                                                                           
Traceback (most recent call last):                                                                                                                          
  File "/home/franzpoeschel/build/nbuild/bin/openpmd-pipe", line 17, in <module>                                                                            
    pipe.main()                                                                                                                                             
  File "/home/franzpoeschel/build/nbuild/lib/python3.11/site-packages/openpmd_api/pipe/__main__.py", line 366, in main                                      
    run_pipe.run()                                                                                                                                          
  File "/home/franzpoeschel/build/nbuild/lib/python3.11/site-packages/openpmd_api/pipe/__main__.py", line 211, in run                                       
    outseries = io.Series(self.outfile, io.Access.create,                                                                                                   
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                   
openpmd_api.openpmd_api_cxx.ErrorWrongAPIUsage: Wrong API usage: Unable to automatically determine filename extension. Please specify in some way.          
[~Series] An error occurred: fileBased output can not be written with no iterations.


> bin/openpmd-pipe --infile ../samples/git-sample/3d-bp4/example-3d-bp4_%T.%E --outfile out.%E --outconfig 'backend = "hdf5"'
Opening data source
Opening data sink
Opened input and output
Iteration 550 contains 9 meshes:
         B
         E
         e_all_chargeDensity
         e_all_energyDensity
         e_all_particleMomentumComponent
         i_all_chargeDensity
         i_all_energyDensity
         i_all_particleMomentumComponent
         picongpu_idProvider

Iteration 550 contains 2 particle species:
[…]


> ls 
out.h5


> bin/openpmd-ls out.%E                                  
openPMD series: out
openPMD standard: 1.1.0
openPMD extensions: 0

data author: unknown
data created: 2021-07-15 13:51:27 +0000 
data backend: HDF5
generating machine: unknown
generating software: PIConGPU (version: 0.6.0-dev)
generating software dependencies: unknown

number of iterations: 2 (groupBased)
  all iterations: 550 600 

number of meshes: 0 # this is a bug

number of particle species: 2
  all particle species:
    e
    i



> bin/openpmd-pipe --infile ../samples/git-sample/3d-bp4/example-3d-bp4_%T.%E --outfile out.%E --outconfig 'backend = "adios2"'
Opening data source
Opening data sink
Opened input and output
Iteration 550 contains 9 meshes:
         B
         E
         e_all_chargeDensity
         e_all_energyDensity
         e_all_particleMomentumComponent
         i_all_chargeDensity
         i_all_energyDensity
         i_all_particleMomentumComponent
         picongpu_idProvider

Iteration 550 contains 2 particle species:



> ls 
out.bp
out.h5



> bin/openpmd-ls out.%E            
An error occurred while opening the specified openPMD series!
Read Error in frontend 
Object type:    File
Error type:     Other
Further description:    Found ambiguous filename extensions on disk: '.h5' and '.bp'.
```

Todo:

- [x] Testing
- [x] Documentation
- [x] (Maybe) Adapt for Linear and Append access types
- [x] Fix (probably unrelated) bug found in openpmd-pipe

### CODE REVIEW

File: include/openPMD/IO/Format.hpp [hunk 1]
```diff
@@ -36,6 +36,7 @@ enum class Format
     ADIOS2_SSC,
     JSON,
     TOML,
+    GENERIC,
```
Comment (franzpoeschel) [inline include/openPMD/IO/Format.hpp:39]: Check if this is still needed

File: src/Series.cpp [hunk 2]
```diff
@@ -71,9 +77,11 @@ namespace
...
     {
         bool isContained{}; //! pattern match successful
-        int padding{}; //! number of zeros used for padding of iteration
-        Series::IterationIndex_t
-            iteration{}; //! iteration found in regex pattern (default: 0)
+        int padding{}; //! number of zeros used for padding of iteration, zero
+                       //! if no padding
+        Series::IterationIndex_t iteration =
+            0; //! iteration found in regex pattern (default: 0)
+        std::optional<std::string> extension;
```
Comment (franzpoeschel) [inline src/Series.cpp:84]: Check if this is still needed

File: include/openPMD/Series.hpp [hunk 3]
```diff
@@ -758,7 +778,26 @@ OPENPMD_private
...
+            m_series->m_deferred_initialization.has_value())
+        {
+            res = runDeferredInitialization();
+        }
+        return res;
+    }
+    AbstractIOHandler const *IOHandler() const
+    {
+        auto res = Attributable::IOHandler();
+        // std::cout << "Returning " << debug(res) << std::endl;
```
Comment (franzpoeschel) [inline include/openPMD/Series.hpp]:
```suggestion
```

File: src/Series.cpp [hunk 4]
```diff
@@ -512,21 +521,21 @@
         std::string const &directory,
         MappingFunction &&mappingFunction)
     {
-        bool isContained;
-        int padding;
-        Series::IterationIndex_t iterationIndex;
+        // bool isContained;
+        // int padding;
+        // Series::IterationIndex_t iterationIndex;
```
[the remark this answers is not in the document]
Comment (franzpoeschel) [reply src/Series.cpp]:
```suggestion
```

File: src/Series.cpp [hunk 5]
```diff
@@ -512,21 +539,21 @@ namespace
...
         std::set<int> paddings;
         if (auxiliary::directory_exists(directory))
         {
             for (auto const &entry : auxiliary::list_directory(directory))
             {
-                std::tie(isContained, padding, iterationIndex) =
-                    isPartOfSeries(entry);
-                if (isContained)
+                // std::tie(isContained, padding, iterationIndex) =
+                // auto [isContained, padding, iterationIndex, extension] =
```
Comment (franzpoeschel) [inline src/Series.cpp]:
```suggestion
```

File: src/Series.cpp [hunk 6]
```diff
@@ -542,25 +569,220 @@ namespace
...
+            throw error::WrongAPIUsage(
+                "Unable to automatically determine filename extension. Please "
+                "specify in some way.");
+        }
+        else
+        {
+            input->filenameExtension = suffix(input->format);
+        }
+    }
+    // auto handler = buildIOHandler(*input, optionsJson, filepath);
```
Comment (franzpoeschel) [inline src/Series.cpp]:
```suggestion
```

File: src/Series.cpp [hunk 7]
```diff
@@ -542,25 +567,240 @@ namespace
...
+                {
+                    extension = ext;
+                }
+            });
+        if (extension.has_value())
+        {
+            if (!additional_extensions.empty())
+            {
+                std::stringstream error;
+                error << "Found inconsistent filename extensions on disk: ";
```
Comment (franzpoeschel) [inline src/Series.cpp]:
```suggestion
                error << "Found ambiguous filename extensions on disk: ";
```

File: include/openPMD/Series.hpp [hunk 8]
```diff
@@ -41,8 +42,10 @@
 #include <cstdint> // uint64_t
 #include <deque>
 #include <map>
```
Comment (ax3l) [inline include/openPMD/Series.hpp:45]:
```suggestion
#include <functional>
#include <map>
```

File: include/openPMD/Series.hpp [hunk 9]
```diff
@@ -41,8 +42,10 @@
...
+#include <memory>
 #include <optional>
 #include <set>
+#include <stdexcept>
 #include <string>
```
Comment (ax3l) [inline include/openPMD/Series.hpp:50]:
```suggestion
#include <string>
#include <tuple>
```

File: include/openPMD/Series.hpp [hunk 10]
```diff
@@ -229,13 +249,50 @@ class Series : public Attributable
 #endif
 
     /**
-     * @brief Construct a new Series
-     *
-     * @param filepath The backend will be determined by the filepath extension.
+     * @brief Construct a new Series.
```
Comment (ax3l) [inline include/openPMD/Series.hpp:252]: Let's add a `.. tip::` in the `First Read` overview:
https://openpmd-api.readthedocs.io/en/latest/usage/firstread.html#open

That explains `%T` and `%E` in 1-2 sentences.

**Review by ax3l — APPROVED** [review]
This is awesome. Please add the `.. tip::` and then we ready.

File: include/openPMD/Series.hpp (hunk 10 above)
Comment (ax3l) [reply include/openPMD/Series.hpp:252]:
```
.. tip::

   Use the wildcards `%T` to match filename encoded iterations and `%E` to match supported extensions such as `.h5` and `.bp`.

**Comment by PrometheusPi** [comment]
Thanks for implementing that file ending wild card @franzpoeschel

**Pull request merged**